File: hardware/scala/nafarr/system/syscon/Syscon.scala
// SPDX-FileCopyrightText: 2026 aesc silicon
//
// SPDX-License-Identifier: CERN-OHL-W-2.0

package nafarr.system.syscon

import spinal.core._
import spinal.lib._
import spinal.lib.bus.misc.BusSlaveFactory
import spinal.lib.bus.amba3.apb._
import spinal.lib.bus.tilelink.{
  Bus => TileLinkBus,
  BusParameter => TileLinkParameter,
  SlaveFactory => TileLinkSlaveFactory
}
import spinal.lib.bus.wishbone._

import nafarr.{IpIdentification, Vendor, Platform, PlatformClass, Product, Feature}
import nafarr.peripherals.PeripheralsComponent

object Syscon {

  case class Parameter(
      vendor: Vendor.E,
      platform: Platform.E,
      platformClass: PlatformClass.E,
      product: Product.E,
      refClockHz: Long,
      siliconMajor: Int = 0,
      siliconMinor: Int = 1,
      features: List[Feature.E] = List(),
      buildDate: Long = Parameter.buildTimestamp
  )
  object Parameter {
    def buildTimestamp: Long =
      sys.env.get("BUILD_DATE").map(_.toLong).getOrElse(System.currentTimeMillis() / 1000)
  }

  object Regs {
    def apply(base: BigInt) = new Regs(base)
  }
  class Regs(base: BigInt) {
    val identity = base + 0x00
    val siliconRev = base + 0x04
    val buildDate = base + 0x08
    val refClock = base + 0x0c
    val featureRegCount = (Feature.elements.size + 31) / 32
    val featureInfo = base + 0x10
    def features(idx: Int) = base + 0x14 + (idx * 4)
  }

  /** All registers are read-only constants derived from compile-time Parameters.
    * No hardware logic is required; the mapper drives the bus directly.
    *
    * identity register layout:
    *   [31:24] = platformClass  (PlatformClass enum ordinal)
    *   [23:16] = product        (Product enum ordinal)
    *   [15:8]  = platform       (Platform enum ordinal)
    *   [7:0]   = vendor         (Vendor enum ordinal)
    *
    * features registers: one 32-bit register per group of 32 Feature ordinals.
    * features(0) covers ordinals 0-31, features(1) covers 32-63, etc.
    * Bit N in features(i) is set when ordinal (i*32 + N) is present in p.features.
    * Register count is derived from the Feature enum size.
    * Populated automatically by the SoC builder via sysconFeatures
    * on each IP's Core class.
    */
  case class Mapper(busCtrl: BusSlaveFactory, p: Parameter) extends Area {
    val idCtrl = IpIdentification(IpIdentification.Ids.Syscon, 1, 0, 0)
    idCtrl.driveFrom(busCtrl)
    val regs = Regs(idCtrl.length)

    // identity: [7:0]=vendor [15:8]=platform [23:16]=product [31:24]=platformClass
    busCtrl.read(
      B(p.platformClass, 8 bits) ## B(p.product, 8 bits) ## B(p.platform, 8 bits) ## B(
        p.vendor,
        8 bits
      ),
      regs.identity
    )

    busCtrl.read(B(p.siliconMajor, 16 bits) ## B(p.siliconMinor, 16 bits), regs.siliconRev)

    busCtrl.read(B(p.buildDate, 32 bits), regs.buildDate)

    busCtrl.read(B(p.refClockHz, 32 bits), regs.refClock)

    busCtrl.read(B(0, 24 bits) ## B(regs.featureRegCount, 8 bits), regs.featureInfo)

    // features: bit N in register i set for Feature with ordinal (i*32 + N)
    var featureMask: BigInt = 0
    p.features.foreach { f => featureMask |= BigInt(1) << f.position }
    for (i <- 0 until regs.featureRegCount) {
      val regBits = (featureMask >> (i * 32)) & ((BigInt(1) << 32) - 1)
      busCtrl.read(B(regBits, 32 bits), regs.features(i))
    }

  }

  class Core[T <: spinal.core.Data with IMasterSlave](
      p: Parameter,
      busType: HardType[T],
      factory: T => BusSlaveFactory
  ) extends PeripheralsComponent {
    val io = new Bundle {
      val bus = slave(busType())
    }
    val busCtrl = factory(io.bus)
    val mapper = Mapper(busCtrl, p)

    override def headerBareMetal(name: String, address: BigInt, size: BigInt) = {
      val baseAddress = "%08x".format(address.toInt)
      s"""#define ${name.toUpperCase}_BASE\t\t0x${baseAddress}\n"""
    }
  }
}

case class Apb3Syscon(
    p: Syscon.Parameter,
    busConfig: Apb3Config = Apb3Config(8, 32)
) extends Syscon.Core[Apb3](
      p,
      Apb3(busConfig),
      Apb3SlaveFactory(_)
    )

case class TileLinkSyscon(
    p: Syscon.Parameter,
    busConfig: TileLinkParameter = TileLinkParameter.simple(8, 32, 4, 1)
) extends Syscon.Core[TileLinkBus](
      p,
      TileLinkBus(busConfig),
      new TileLinkSlaveFactory(_, false)
    )

case class WishboneSyscon(
    p: Syscon.Parameter,
    busConfig: WishboneConfig = WishboneConfig(8, 32)
) extends Syscon.Core[Wishbone](
      p,
      Wishbone(busConfig),
      WishboneSlaveFactory(_)
    )

File: hardware/scala/nafarr/system/syscon/Syscon.cs
// SPDX-FileCopyrightText: 2026 aesc silicon
//
// SPDX-License-Identifier: CERN-OHL-W-2.0
//
// Renode model of the Nafarr syscon (nafarr.system.syscon.Syscon). All registers are read-only
// constants (IpIdentification + identity/siliconRev/buildDate/refClock/featureInfo/features),
// mirroring Syscon.scala's Mapper. Values are constructor parameters so each SoC can supply its own.
//
// Register map (offsets from base; Regs base = IpIdentification.length = 8):
//   0x00 header     (RO)  IpIdentification
//   0x04 version    (RO)
//   0x08 identity   (RO)  platformClass[31:24] | product[23:16] | platform[15:8] | vendor[7:0]
//   0x0C siliconRev (RO)  siliconMajor[31:16] | siliconMinor[15:0]
//   0x10 buildDate  (RO)
//   0x14 refClock   (RO)  reference clock in Hz
//   0x18 featureInfo(RO)  featureRegCount[7:0]
//   0x1C + i*4 features(i) (RO)  feature bitmask, 32 ordinals per register
//
using Antmicro.Renode.Core;
using Antmicro.Renode.Logging;
using Antmicro.Renode.Peripherals.Bus;

namespace Antmicro.Renode.Peripherals.Miscellaneous
{
    public class Syscon : IDoubleWordPeripheral, IKnownSize
    {
        public Syscon(IMachine machine, uint refClockHz = 50000000, uint identity = 0,
                      uint siliconMajor = 0, uint siliconMinor = 0, uint buildDate = 0,
                      ulong featureMask = 0)
        {
            this.refClockHz = refClockHz;
            this.identity = identity;
            this.siliconRev = (siliconMajor << 16) | (siliconMinor & 0xFFFF);
            this.buildDate = buildDate;
            this.featureMask = featureMask;
            this.featureRegCount = featureMask == 0 ? 1 : (BitLength(featureMask) + 31) / 32;
        }

        public void Reset()
        {
        }

        public uint ReadDoubleWord(long offset)
        {
            switch(offset)
            {
                case HeaderOffset:     return (uint)(((Api & 0xFF) << 24) | ((Length & 0xFF) << 16) | (Id & 0xFFFF));
                case VersionOffset:    return 0x01000000; // 1.0.0
                case IdentityOffset:   return identity;
                case SiliconRevOffset: return siliconRev;
                case BuildDateOffset:  return buildDate;
                case RefClockOffset:   return refClockHz;
                case FeatureInfoOffset:return featureRegCount & 0xFF;
            }
            if(offset >= FeaturesOffset && offset < FeaturesOffset + featureRegCount * 4)
            {
                var idx = (int)((offset - FeaturesOffset) / 4);
                return (uint)((featureMask >> (idx * 32)) & 0xFFFFFFFF);
            }
            this.Log(LogLevel.Warning, "Unhandled read at offset 0x{0:X}", offset);
            return 0;
        }

        public void WriteDoubleWord(long offset, uint value)
        {
            this.Log(LogLevel.Warning, "Write to read-only syscon register at offset 0x{0:X}", offset);
        }

        public long Size => 0x1000;

        private static uint BitLength(ulong v)
        {
            uint n = 0;
            while(v != 0) { n++; v >>= 1; }
            return n;
        }

        private readonly uint refClockHz;
        private readonly uint identity;
        private readonly uint siliconRev;
        private readonly uint buildDate;
        private readonly ulong featureMask;
        private readonly uint featureRegCount;

        // IpIdentification header constants (Syscon.scala: id = Syscon = 24, length = 8).
        private const int Api = 0;
        private const int Length = 8;
        private const int Id = 24;

        private const long HeaderOffset = 0x00;
        private const long VersionOffset = 0x04;
        private const long IdentityOffset = 0x08;
        private const long SiliconRevOffset = 0x0C;
        private const long BuildDateOffset = 0x10;
        private const long RefClockOffset = 0x14;
        private const long FeatureInfoOffset = 0x18;
        private const long FeaturesOffset = 0x1C;
    }
}
